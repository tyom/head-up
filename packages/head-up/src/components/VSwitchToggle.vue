<template>
  <div class="SwitchToggle">
    <div :class="{_toggled: toggled}" class="toggle-indicator">
      <div class="toggle-head"/>
      <input
        :id="inputId"
        :checked="toggled"
        type="checkbox"
        @input="toggle"
      >
    </div>
    <label
      v-if="label"
      :for="inputId"
      v-text="label"
    />
  </div>
</template>

<script>
export default {
  props: {
    label: {
      type: String,
      default: '',
    },
    value: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      toggled: this.value,
    }
  },
  computed: {
    inputId() {
      return `switch-toggle-${this._uid}`
    },
  },
  methods: {
    toggle(evt) {
      this.toggled = evt.target.checked
      this.$emit('input', this.toggled)
    },
  },
}
</script>

<style scoped>
:root {
  --toggle-width: 2em;
  --toggle-height: 1em;
}

.SwitchToggle {
  display: flex;
  align-items: center;
  line-height: 1;
}

.toggle-indicator {
  position: relative;
  background-color: #bbb;
  border: 2px solid #0004;
  padding: 1px;
  border-radius: var(--toggle-height);
  width: var(--toggle-width);
  margin-right: 0.5em;
  transition: 0.2s;
}

.toggle-head {
  transition: 0.2s;
  border-radius: 50%;
  background-color: #fff;
  border: 1px solid #0005;
  height: var(--toggle-height);
  width: var(--toggle-height);
}

._toggled {
  &.toggle-indicator {
    background-color: #95de85;
  }

  & .toggle-head {
    background-color: #fff;
    transform: translateX(calc(100% - 6px));
  }
}

input {
  font-size: 1em;
  opacity: 0;
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
}
</style>
