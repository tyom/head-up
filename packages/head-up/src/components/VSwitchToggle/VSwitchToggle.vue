<template>
  <div class="SwitchToggle">
    <div :class="{_toggled: toggled}" class="toggle-indicator">
      <div class="toggle-head"/>
      <input
        v-model="toggled"
        :id="inputId"
        :name="name"
        type="checkbox"
      >
    </div>
    <label
      v-if="label"
      :for="inputId"
      v-text="label"
    />
  </div>
</template>

<script>
export default {
  props: {
    label: {
      type: String,
      required: true,
    },
    name: {
      type: String,
      default: null,
    },
    value: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      toggled: this.value,
    };
  },
  computed: {
    inputId() {
      return `switch-toggle-${this._uid}`;
    },
  },
  watch: {
    toggled(newVal) {
      this.$emit('input', newVal);
    },
  },
};
</script>

<style scoped>
:root {
  --toggle-width: 2em;
  --toggle-height: 1em;
}

.SwitchToggle {
  display: flex;
  align-items: center;
  line-height: 1;
}

.toggle-indicator {
  position: relative;
  background-color: #bbb;
  border: 2px solid #0004;
  box-shadow: inset 1px 1px rgba(#000, 0.1);
  padding: 1px;
  border-radius: var(--toggle-height);
  width: var(--toggle-width);
  margin-right: 0.5em;
  transition: 0.2s;
}

.toggle-head {
  transition: 0.2s;
  border-radius: 50%;
  background-color: #fff;
  border: 1px solid rgba(#000, 0.2);
  box-shadow: 1px 1px rgba(#000, 0.06), inset -1px -1px 1px rgba(#000, 0.1);
  background-clip: padding-box;
  height: var(--toggle-height);
  width: var(--toggle-height);
}

._toggled {
  &.toggle-indicator {
    background-color: #95de85;
  }

  & .toggle-head {
    background-color: #fff;
    transform: translateX(calc(100% - 6px));
  }
}

input {
  font-size: 1em;
  opacity: 0;
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
}
</style>
