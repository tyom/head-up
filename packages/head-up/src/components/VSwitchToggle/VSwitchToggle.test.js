import { shallowMount } from '@vue/test-utils';
import VSwitchToggle from './VSwitchToggle';

test('render default', () => {
  const wrapper = shallowMount(VSwitchToggle, {
    propsData: {
      label: 'Check it',
    },
  });
  expect(wrapper).toMatchSnapshot();
});

test('change on input click', () => {
  const wrapper = shallowMount(VSwitchToggle, {
    propsData: {
      label: 'Check it',
    },
  });

  const input = wrapper.find('input');

  expect(wrapper.vm.toggled).toEqual(false);

  input.trigger('click');

  expect(wrapper.vm.toggled).toEqual(true);
});

test('change on label click', () => {
  const wrapper = shallowMount(VSwitchToggle, {
    propsData: {
      label: 'Check it',
    },
  });

  const label = wrapper.find('label');

  expect(wrapper.vm.toggled).toEqual(false);

  label.trigger('click');

  expect(wrapper.vm.toggled).toEqual(true);
});

test('emits input event', () => {
  const wrapper = shallowMount(VSwitchToggle, {
    propsData: {
      label: 'Check it',
    },
  });

  const input = wrapper.find('input');
  input.trigger('click');

  expect(wrapper.emitted('input')[0]).toEqual([true]);
});
