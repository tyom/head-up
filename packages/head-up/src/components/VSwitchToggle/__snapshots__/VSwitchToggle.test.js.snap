// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render default 1`] = `
<div class="SwitchToggle">
  <div class="toggle-indicator">
    <div class="toggle-head"></div> <input id="switch-toggle-1" type="checkbox">
  </div> <label for="switch-toggle-1">Check it</label>
</div>
`;
